internal/cli/cli.go

```diff
@@ -1,35 +1,126 @@
 package main
 
 import (
+	"errors"
+	"flag"
 	"fmt"
+	"io"
 	"log"
 	"os"
 	"time"
 
 	"github.com/innosat-mats/rac-extract-payload/internal/common"
 )
 
-func callback(pkg common.DataRecord) {
-	fmt.Println(pkg)
+func generateDiskWriterCallback(
+	output string,
+	writeImages bool,
+	writeTimeseries bool,
+) common.ExtractCallback {
+	return func(pkg common.DataRecord) {
+		// This is just a placeholder, should write to output directory
+	}
 }
 
-func main() {
-	batch := make([]common.StreamBatch, len(os.Args)-1)
-	for n, filename := range os.Args[1:] {
+func generateStdoutCallback(
+	out io.Writer,
+	writeTimeseries bool,
+) common.ExtractCallback {
+
+	return func(pkg common.DataRecord) {
+		if writeTimeseries {
+			fmt.Fprintf(out, "%+v\n", pkg)
+		}
+	}
+}
+
+var skipImages *bool
+var skipTimeseries *bool
+var outputDirectory *string
+var stdout *bool
+
+//myUsage replaces default usage since it doesn't include information on non-flags
+func myUsage() {
+	fmt.Println("Extracts information from Innosat-MATS rac-files")
+	fmt.Println()
+	fmt.Printf("Usage: %s [OPTIONS] rac-file ...\n", os.Args[0])
+	fmt.Println()
+	flag.PrintDefaults()
+}
+
+func getCallback(
+	stdout bool,
+	outputDirectory string,
+	skipImages bool,
+	skipTimeseries bool,
+) (common.ExtractCallback, error) {
+	if outputDirectory == "" && !stdout {
+		flag.Usage()
+		fmt.Println("\nExpected an output directory")
+		return nil, errors.New("Invalid arguments")
+	}
+	if skipTimeseries && (skipImages || stdout) {
+		fmt.Println("Nothing will be extracted, only validating integrity of rac-file(s)")
+	}
+
+	if stdout {
+		return generateStdoutCallback(os.Stdout, !skipTimeseries), nil
+	}
+	return generateDiskWriterCallback(
+		outputDirectory,
+		!skipImages,
+		!skipTimeseries,
+	), nil
+}
+
+func processFiles(
+	extractor common.ExtractFunction,
+	inputFiles []string,
+	callback common.ExtractCallback,
+) error {
+	batch := make([]common.StreamBatch, len(inputFiles))
+	for n, filename := range inputFiles {
 		f, err := os.Open(filename)
 		defer f.Close()
 		if err != nil {
-			log.Fatalln(err)
+			return err
 		}
 
 		batch[n] = common.StreamBatch{
 			Buf: f,
 			Origin: common.OriginDescription{
-				Name: filename,
-				Date: time.Now(),
+				Name:           filename,
+				ProcessingDate: time.Now(),
 			},
 		}
 
 	}
-	common.ExtractData(callback, batch...)
+	extractor(callback, batch...)
+	return nil
+}
+
+func init() {
+	skipImages = flag.Bool("skip-images", false, "Extract images from rac-files.\n(Default: false)")
+	skipTimeseries = flag.Bool("skip-timeseries", false, "Extract timeseries from rac-files.\n(Default: false)")
+	outputDirectory = flag.String("output", "", "Directory to place images and/or timeseries data")
```

Comment:
Does "" as default mean that it is mandatory? If so, then better as non-flag argument. Otherwise, make it clear what the default is ("./"?).

Reply:
No it doesn't make it mandatory. And it's not, only mandatory if `-stdout` is not included. The default value ("") is just a hidden internal value to conform with interfaces of the `flag` lib and type restrictions and won't be accepted.

Reply:
I find the UX better to have to explicitly state the output directory so one does not accidentally write loads of files into e.g. the directory one happened to execute from.